### PR TITLE
JSON output

### PR DESCRIPTION
**Pull request by robrix**
I'm using the CLI tool instead of linking against the library, and I want JSON output instead of writing a parser for the current output format. My goal with this PR is to define something that will work for my use case first, and then refine that into a stable schema ideally in a  future PR.

This is the first rust I've written since 0.4, around 10y ago if I recall correctly. Please bear that in mind, and be excruciatingly picky with code review—as long as you explain the feedback 😊

### CODE REVIEW

File: src/bin/tree-sitter-graph/main.rs
Comment (robrix) [inline src/bin/tree-sitter-graph/main.rs:69]: @patrickt wisely suggested that we branch here rather than try to shoehorn JSON formatting into the `display_with` machinery.

File: src/graph.rs
Comment (robrix) [inline src/graph.rs:97]: This is really complicating (not to mention bloating) this file, so I'm likely to move it elsewhere.

File: src/graph.rs
Comment (robrix) [inline src/graph.rs:103]: The `display_with` machinery works by creating a new ~~typecla~~—ahem, **_trait_**—which provides some context data to pass around for resolving interned strings. We require the same context for the same reason, but it was most convenient to follow the approach taken by `display_with`, above (but folded in this PR view): define a struct type locally and give it an implementation of `serde::Serialize`.

Because a `Serializer`'s methods call into its own API, however, we can't reasonably distribute the construction of these `serialize` implementations around all of the types in question behind some convenient (and useful in isolation) method like `display_with`. For example, calling `serialize_seq` returns a serializer with a `serialize_element` method which takes values with a `Serialize` trait—so it only allows for top-down folds, not bottom-up. We could still provide a `serialize_with` method for each of the public types which, constructing the corresponding private JSON struct and calling into it. But we wouldn't be able to recur via that method.

Thus we're somewhat constrained in the precise pattern we employ (and thus in how easy it will be to test this monstrosity), but at least we don't have to keep all of these types inside this one method 😄

File: src/graph.rs
Comment (robrix) [inline src/graph.rs:128]: It's tempting to use a struct given that the shape of the map in question is static, but serializers would then bloat how the values are encoded with their type name despite their only ever occurring in a homogeneously typed position.

File: src/graph.rs
Comment (robrix) [inline src/graph.rs:137]: It was this or make another inner struct just for vectors of edges. Tragically, there is no `Serialize` implementation for `SmallVec`.

Note that `collect_seq` is not applicable since `serialize` returns a `Result` of unit and not, say, something which you could pass in as a `Serialize`-implementing value for the second argument to `serialize_entry` here. (I take issue with several points of `serde`'s design.)

File: src/graph.rs
Comment (robrix) [inline src/graph.rs:140]: Rust really needs `do`-notation.

File: src/graph.rs
Comment (robrix) [inline src/graph.rs:202]: I think I probably want to use one of the variant serialization options to handle this; it will also catch the various integer cases (below).

File: src/graph.rs
Comment (robrix) [inline src/graph.rs:217]: I think this should probably be a `Serialize` implementation for `Graph`, but in the meantime we can get readable JSON out of it.

**Review by robrix — COMMENTED** [review]
Ready for review!

File: src/graph.rs
Comment (hendrikvanantwerpen) [reply src/graph.rs:128]: That seems the right way, and also more idomatic JSON as a result.

File: src/graph.rs
Comment (hendrikvanantwerpen) [inline src/graph.rs]: Out of curiosity, is `where S` simply a way to get local abbreviation here, or does this do anything else?

File: src/graph.rs
Comment (hendrikvanantwerpen) [inline src/graph.rs:131]: I wonder if it is better always add the field, even if it is empty. It makes it easier for consumers, because there are fewer cases to handle.

**Review by hendrikvanantwerpen — COMMENTED** [review]
Looks nice and sweet so far :) I've added some small comments below.

File: src/graph.rs
Comment (patrickt) [reply src/graph.rs]: In Rust you express trait constraints not directly within the signature, but in a `where` clause afterwards. I’m not sure if this is a hard language-level requirement or just a general practice, but you see it all over the stdlib.

File: src/graph.rs
Comment (patrickt) [reply src/graph.rs:131]: Yeah, I think uniformity of JSON output trumps the extremely minor expense of serializing an empty array.

Comment (robrix) [reply src/graph.rs:131]: Good points. I was optimizing for human-readability, but that's really not our intended use case.

File: src/graph.rs
Comment (hendrikvanantwerpen) [inline src/graph.rs:146]: Is `u32` a `GraphNodeID`?

File: src/graph.rs
Comment (hendrikvanantwerpen) [inline src/graph.rs:97]: This works out really nicely. I should use that in other places...

File: src/graph.rs
Comment (hendrikvanantwerpen) [inline src/graph.rs:120]: Are the spaces before the dots intentional here?

File: src/graph.rs
Comment (hendrikvanantwerpen) [inline src/graph.rs:108]: Not sure if this is a good idea, but you could implement a method on `Context` to create these `InContext` values, so that you could write `ctx.with(&(node_index, node))`.

File: src/graph.rs
Comment (hendrikvanantwerpen) [inline src/graph.rs:124]: This corresponds exactly to the position of this entry in the enclosing list, right? Is it useful to have this explicitly as a property as well, you think? Or do you consider it purely accidental that this index happens to coincide with the position, and you should always search the list for the entry with the correct index (which in that case should perhaps be named `id`)?

File: src/graph.rs
Comment (hendrikvanantwerpen) [reply src/graph.rs:200]: I assume that would produce some form of tagged output? I think that makes sense for values.

File: src/graph.rs
Comment (hendrikvanantwerpen) [inline src/graph.rs:207]: Do syntax node IDs have much meaning in the output? Not that I can think of a better thing to output here...

**Review by hendrikvanantwerpen — COMMENTED** [review]
I left a few more questions, some nits and some on the output format.

File: src/graph.rs
Comment (patrickt) [reply src/graph.rs:120]: I think so, as otherwise the lexer would try to recognize it as a float. You can also do `(self.0).1` but it works without the parens if you add a space.

File: src/graph.rs
Comment (patrickt) [reply src/graph.rs:96]: @robrix I think I left that thumbs down accidentally, I have no strong feelings

File: src/graph.rs
Comment (hendrikvanantwerpen) [reply src/graph.rs:120]: Oh, wow

File: src/graph.rs
Comment (robrix) [reply src/graph.rs:146]: Aha, so it is! Good catch.

File: src/graph.rs
Comment (robrix) [reply src/graph.rs:120]: The autoformatter inserts these spaces on my behalf, FWIW.

File: src/graph.rs
Comment (robrix) [reply src/graph.rs:97]: Yeah, I was quite pleased with it.

File: src/graph.rs
Comment (robrix) [reply src/graph.rs:108]: I dig it 👍🏻

File: src/graph.rs
Comment (robrix) [reply src/graph.rs:124]: AFAIK it does correspond thus, but I'm considering this as future-proofing against that decision changing. `display_with` makes the same decision, which is probably what I was cribbing  from.

`id` is definitely a better name, and probably something similar applies to `Edge`s.

File: src/graph.rs
Comment (robrix) [reply src/graph.rs:200]: Yeah, that's what I'm thinking. I'd like to merge this without those for the moment if folks are cool with that (what's the release process like on this repo?); it's not that this change will take a long time, more that it will _definitely_ interfere with work on the `semantic-analysis` side, and I'd like to postpone that interference for a week or so.

File: src/graph.rs
Comment (robrix) [reply src/graph.rs:207]: I can't imagine so, unless you have the syntax tree as well (which you don't). I was thinking about this, and maybe we could get a rough summary of the node: its rule name, its range, etc. Again, I'd like to punt on that for just a little while if possible.

**Review by robrix — COMMENTED** [review]
Applied some more feedback from @hendrikvanantwerpen (🙇🏻), and this is now ready for re-review. As noted in the comments, I'd ideally like to merge more or less as-is and address the FIXMEs in a future PR to avoid holding up work in `semantic-analysis`, but I'll leave that to reviewers' discretion.

In the meantime, I've opted not to refactor the implementation into a separate file and/or a `Serialize` trait implementation for `Graph` for the time being, mostly so that I can focus my efforts on higher priority tasks. Further to that, it's not immediately obvious to me which of the various factoring options are actually valuable, vs. busywork. Again, input is welcome, but I'd strongly prefer to punt this particular task for now.

**Review by patrickt — APPROVED** [review]
Looks beautiful ✨

File: Cargo.toml
Comment (dcreager) [inline Cargo.toml:25]: Is there anything in these particular patch versions that you're depending on, or could you list the requirement as `1.0` in both cases?

File: src/bin/tree-sitter-graph/main.rs
Comment (dcreager) [reply src/bin/tree-sitter-graph/main.rs:69]: :100:  Hard agree

File: src/graph.rs
Comment (dcreager) [reply src/graph.rs:126]: If the autoformatter says this is the way to go, this is the way to go :smile:

File: src/graph.rs
Comment (dcreager) [reply src/graph.rs:128]: I believe serde has a way to "flatten" a struct (or maybe the field of a struct?) to remove that extra level of naming, but I don't suggest blocking this PR on researching and changing to that.

File: src/graph.rs
Comment (dcreager) [reply src/graph.rs:206]: > what's the release process like on this repo?

ad hoc, need to write a `script` for it.  Nothing automatic, so it's fine to have this merged and unreleased for a period of time.

File: src/graph.rs
Comment (dcreager) [reply src/graph.rs:213]: :+1: to punt.  I do think it's worth including the syntax tree (possibly with an option to give you (a) the whole tree, or (b) just the nodes mentioned in the output graph), but that doesn't have to block this PR.

**Review by dcreager — APPROVED** [review]
:+1: — my only open question is about whether we can loosen the new dependencies in `Cargo.toml`

File: Cargo.toml
Comment (robrix) [reply Cargo.toml:25]: Looks like they both loosen just fine 👍🏻

File: src/graph.rs
Comment (robrix) [reply src/graph.rs:213]: Great idea 👍🏻

**Pull request merged**
